### ISSUE TITLE
Resize Lab: Implement transition from Visual Dimensions  to Actual Dimensions 

### ISSUE THREAD

**Issue by Leook0209** [PowerPointLabs/PowerPointLabs#1069]
When using the ADJUST SLIGHTLY feature on a slanted shape, switching the setting from 'Use Actual Height / Width' to 'Use Visual Dimensions' works fine. But switching from 'Use Visual Dimensions' to 'Use Actual Height / Width' does not work. It still uses visual dimensions instead of actual.

**Comment by IanTeo** [comment]
May I do this? I think I may have an idea on how to solve it

**Comment by damithc** [comment]
@IanTeo sure give it a try

**Comment by IanTeo** [comment]
Before Resize: 
![beforeresize](https://cloud.githubusercontent.com/assets/12381148/22145066/cff31908-df3a-11e6-9ca3-1c1091b21ecc.png)

After Resize:
![afterresize](https://cloud.githubusercontent.com/assets/12381148/22145069/d12fe58a-df3a-11e6-8522-96f2a390ef97.png)

I believe that the bug is caused by the fact that resizing the visual size of the box also causes the boundary of the box to be redrawn, as can be seen from the 2 images. Once the boundary has been redrawn, the new boundary is treated as the actual size, leading to the bug.

Ideally, the boundary box should not be redrawn when calculating the size. Of course, there must be a reason why the programmer who did this portion did it like this (Maybe it was not possible to do it without changing the boundary box?). If it is possible, could the person who made these codes explain to me why it has to be this way? I can't figure it out.

My solution to this problem without changing the structure of the codes would be a workaround.
- The shapes that need to be resized will be duplicated
- Calculation of new size will be done on the duplicated shapes
- Apply the new dimensions to the original shapes
- Delete the duplicated shapes

Lastly, the current unit tests show that this redrawing of the boundary box is intentional. @damithc should I change the tests to fit the changes that I will make?

**Comment by damithc** [comment]
@youthinkk @yancmo any input?

**Comment by yancmo** [comment]
Yes, it was sort of intentional, which is a drawback when we use free-form shapes. We decided to use this because we found it very hard to calculate the dimensions to resize for rotated objects. This is largely because the .left and .top properties correspond to the top left point of the bounding box, **not the actual shape itself** irregardless of whether it is rotated. Things get even more complicated when you try to resize non-rectangle and complex shapes, as it is difficult to get the "true" left/top/right/bottom-most points of the shapes.
Of course, feel free to give it a try :). On a side note, why do you need to duplicate a shape before applying the calculated dimensions?

**Comment by IanTeo** [comment]
Because calculating the dimensions will cause the shape's boundary to change, which causes this bug. So what I proposed to do was to create a duplicate, do the calculations on the duplicated shape, apply the width/height to the original shape, then delete the duplicated shape it. This way, the original shape will have it's dimensions changed accordingly and still retain it's original boundary (Instead of turning the boundary into a upright square surrounding the object).

**Comment by IanTeo** [comment]
So to clarify, the current action of visual resize is: **Resize according to what the user wants, but change the boundary of the shape** (Drawback of using free-form shapes)

But ideally, **we do not want to change the boundary of the shapes that the user created**, right?

**Comment by youthinkk** [comment]
Yes, you are right! The current version resize according to what the user wants **visually**, but also change the boundary of the shape. We did so due to the following limitations of the default PowerPoint shape properties:
1. API of shapes is bound to the bounding box, not the shape itself
2. Only the height and weight of the bounding box can be adjusted with regards to the size
3. The bounding box is only in square or rectangle

By using your method, there are some possible problems that I can think of:
* From (2), it's unlikely you can adjust the size for irregular shapes mentioned by @yancmo 
* New height and width on the duplicated freeform bounding box cannot be applied onto original rotated shape since there is no information of the nodes in the duplicated shape for you to calculate. 

However, if you can access the internal nodes and lines of the shape, the problem should be solved easily. We did not manage to find available API during our development that time. 😝

**Comment by IanTeo** [comment]
Set of Shapes:
![image](https://cloud.githubusercontent.com/assets/12381148/22198671/77d0c928-e192-11e6-85e4-b84360aad642.png)

Current Behavior:
![image](https://cloud.githubusercontent.com/assets/12381148/22198166/3f8d1190-e190-11e6-9569-a008968ca248.png)

New Behavior:
![image](https://cloud.githubusercontent.com/assets/12381148/22198090/f32eedb4-e18f-11e6-9d85-8f7d69d11308.png)

My previous solution does not work as @youthinkk and @yancmo mentioned, I could not find an API to access the internal nodes. I tried another solution, which was to adjust the boundary box to the same angle as the original shapes. However, as you can see from the third image, the shape completely changes it's dimensions in some instances, leading the original box angle to be a bad fit for the shape. Thus, this solution does not work too.

Considering that the shape is able to change it's dimensions so drastically when using the resize lab, the optimal solution may actually be to leave it as it is (second image). I cannot think of any way to determine the true new boundary of the box. Maybe someone with an idea could share it with me?

I am able to keep the boundary box of the reference shape (Which is not being modified) the same as the original boundary box (image 3). That way, when the user is switching from visual to actual with the same shape, they still can resize them properly using equalize. **My proposal is to keep the boundary for the reference shape only. That way, the problem is "solved" unless the user wants to change the reference shape.** What do you guys think?

**Comment by yancmo** [comment]
Yup, I think that would be a great enhancement :)
